index.bs

```diff
@@ -133,50 +133,93 @@ run the following steps:
 
 ## Permissions ## {#permissions}
 
-Issue: Figure out some way to integrate this better with the permissions API [[permissions]],
-for example by making the entry be a part of the PermissionDescriptor.
+The <dfn for=PermissionName enum-value>"file-system"</dfn> [=powerful feature=]'s
+permission-related algorithms and types are defined as follows:
+
+: [=permission descriptor type=]
+:: {{FileSystemPermissionDescriptor}}, defined as:
+  <xmp class=idl>
+  enum FileSystemPermissionMode {
+    "read",
+    "readwrite"
+  };
+
+  dictionary FileSystemPermissionDescriptor : PermissionDescriptor {
+    required FileSystemHandle handle;
+    FileSystemPermissionMode mode = "read";
+  };
+  </xmp>
+
+: [=permission state constraints=]
+:: <div algorithm="permission state constraints">
+  To determine [=permission state constraints=] for a {{FileSystemPermissionDescriptor}} |desc|,
+  run these steps:
+
+  1. Let |entry| be |desc|.{{FileSystemPermissionDescriptor/handle}}'s [=FileSystemHandle/entry=].
+  1. If |entry| represents an [=/entry=] in an [=origin private file system=],
+    this descriptor's [=permission state=] should always be {{PermissionState/"granted"}}.
```

Comment:
This block has a lot of "should" in it, which seems unusual for an algorithm.

Does "should" signal a place where implementations can deviate (RFC-style) or is it an assertion? If it's the former, WDYT about something like "To determine [=permission state constraints=] for a {{FileSystemPermissionDescriptor}} |desc|, these steps **should** be run:" followed by the usual style for writing steps?

Reply:
phrasing this as an algorithm felt weird to begin with. It's not really invoked as an algorithm. But it seemed at least the clearest way of expressing the relevant constraints. Probably I should just change the shoulds to musts.

Reply:
Good idea! Thank you!